### PR TITLE
Add Saml binding validator to acs endpoint

### PR DESCRIPTION
**Pull request by pablothedude**
This is done in order to make sure that the 'Current authentication
procedure exceptions' are split up. So the information returned
should be much more informative.

### CODE REVIEW

**Comment by pablothedude** [comment]
https://www.pivotaltracker.com/story/show/164238335

File: src/OpenConext/EngineBlock/Validator/SamlBindingValidator.php [hunk 1]
```diff
@@ -0,0 +1,59 @@
+<?php
```
Comment (MKodde) [inline src/OpenConext/EngineBlock/Validator/SamlBindingValidator.php]: I like the fact that you seperated the binding and request validations into separate validators.

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php [hunk 2]
```diff
@@ -91,8 +91,11 @@ public function authenticatedAt($requestId, Entity $identityProvider)
 
         if ($currentRequest === null) {
             throw new LogicException(
-                'Current authentication procedure cannot be authenticated:'
-                 . ' authentication procedure has not been started'
+                sprintf(
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: The exception messages do not sit right with me, for example: 

> Requested authentication procedure cannot be found:  authentication procedure could not be found for given requestId (%s)

Can be shortened as both sentences now almost convey the same message (redundancy). I also suggest adding more information, to make the exception more easily understandable.

For example:

> The requested authentication procedure with requestId (%s) couldn't be found in the session storage.

Comment (MKodde) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: The same applies for the two other exception texts in this file.

File: src/OpenConext/EngineBlockBundle/Controller/IdentityProviderController.php [hunk 3]
```diff
@@ -115,13 +119,16 @@ public function singleSignOnAction(Request $request, $keyId = null, $idpHash = n
...
+     * @param Request $request
      * @param null|string $keyId
      * @param null|string $idpHash
      * @return \Symfony\Component\HttpFoundation\RedirectResponse|\Symfony\Component\HttpFoundation\Response
-     * @throws Exception
      */
-    public function unsolicitedSingleSignOnAction($keyId = null, $idpHash = null)
+    public function unsolicitedSingleSignOnAction(Request $request, $keyId = null, $idpHash = null)
     {
+        $this->requestValidator->isValid($request);
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Controller/IdentityProviderController.php]: Thanks for fixing this!

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature [hunk 4]
```diff
@@ -278,16 +278,16 @@ Feature:
...
+     And I should see ART code "35954"
 
   Scenario: The SP uses the wrong request parameter while using HTTP Post binding
    Given the SP "Dummy SP" sends a malformed AuthNRequest
      And the SP "Dummy SP" uses the HTTP POST Binding
     When I log in at "Dummy SP"
      And I pass through the SP
-    Then I should see "No SAMLRequest parameter was found in the HTTP \"POST\" request parameters"
-     And I should see ART code "18993"
+    Then I should see "The parameter \"SAMLRequest\" is missing on this SAML SSO endpoint"
```
Comment (MKodde) [inline src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature]: Where you not able to add Behat scenarios for SAMLResponse validation violations. I really value the addition of those tests!

File: tests/unit/OpenConext/EngineBlock/Validator/AcsRequestValidatorTest.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace OpenConext\EngineBlock\Validator;
+
+use OpenConext\EngineBlock\Exception\InvalidRequestMethodException;
+use PHPUnit_Framework_TestCase as TestCase;
+use Symfony\Component\HttpFoundation\Request;
+
+class AcsRequestValidatorTest extends TestCase
```
Comment (MKodde) [inline tests/unit/OpenConext/EngineBlock/Validator/AcsRequestValidatorTest.php]: Throwing of the `MissingParameterException` is not explicitly tested, also goes for the SSO request validator.

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php (hunk 2 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: I'll wait on the outcome of the discussion in Pivotal.

File: src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature (hunk 4 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlockFunctionalTestingBundle/Features/ClearErrorMessages.feature]: I agree, I shall add this.

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php (hunk 2 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: Changed the texts.

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php [hunk 6]
```diff
@@ -110,15 +112,19 @@ public function completeCurrentProcedure($requestId)
         $currentRequest = $this->authenticationProcedures->find($requestId);
         if ($currentRequest === null) {
             throw new LogicException(
-                'Current authentication procedure cannot be completed:'
-                . ' authentication procedure has not been started'
+                sprintf(
+                    'The requested authentication procedure with requestId (%s) couldn\'t be found in the session storage.',
```
Comment (MKodde) [inline src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: This is not what actually happened right? The authn procedure could not be started, this needs fixing. Ditto for the exception message on line 124

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php (hunk 2 above)
Comment (MKodde) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: Thanks for the update, see my new feedback below.

File: src/OpenConext/EngineBlock/Validator/SsoRequestValidator.php [hunk 7]
```diff
@@ -50,21 +46,14 @@ public function isValid(Request $request)
...
-        }
-        if (!($binding instanceof HTTPRedirect || $binding instanceof HTTPPost)) {
-            // We only support HTTP Redirect binding
-            throw new InvalidBindingException(
-                sprintf('The binding type "%s" is not supported on this SAML SSO endpoint', get_class($binding))
+
+        if (($requestMethod ===  Request::METHOD_POST && !$request->request->has('SAMLRequest')) ||
+            ($requestMethod ===  Request::METHOD_GET && !$request->query->has('SAMLRequest'))) {
+            throw new MissingParameterException(
+                sprintf('The parameter "SAMLRequest" is missing on this SAML SSO endpoint')
```
Comment (MKodde) [inline src/OpenConext/EngineBlock/Validator/SsoRequestValidator.php]: This new exception message is not correct in my opinion. An endpoint can not have a missing request parameter right? The request that was fired at the endpoint is missing the parameter. My suggestion is to reinstate the previous message. Also goes for the Acs validator counterpart.

File: src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php (hunk 6 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: This is the mesasge as suggested in:
https://github.com/OpenConext/OpenConext-engineblock/pull/665#discussion_r273446033

Comment (MKodde) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: Yes, the proposal was only for the first exception message. The other two need a similar revision in my opinion, but the copy-paste is not correct and make the error messages incorrect.

Comment (pablothedude) [reply src/OpenConext/EngineBlockBundle/Authentication/AuthenticationState.php]: Done

File: src/OpenConext/EngineBlock/Validator/SsoRequestValidator.php (hunk 7 above)
Comment (pablothedude) [reply src/OpenConext/EngineBlock/Validator/SsoRequestValidator.php]: Changed the message

**Pull request merged**